File: Scala/graph_algorithms/src/main/scala/minimum_spanning_trees/CheapestRoadNetwork.scala
package minimum_spanning_trees

object CheapestRoadNetwork {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Map => MutableMap}
  import scala.collection.breakOut

  final case class Point(x: Double, y: Double)

  private def convertToDoubleList(line: String): List[Double] = line.split(" ").map(_.toDouble).toList

  private def calcDistance(p: Point, q: Point): Double = math.hypot(p.x - q.x, p.y - q.y)

  def calcMinimalSpanningTree(points: List[Point]): Double = {
    val nodesWithCost: MutableMap[Point, Double] = (for { p <- points } yield (p, Double.PositiveInfinity))(breakOut)
    val startingPoint: Point = points.head
    nodesWithCost(startingPoint) = 0.0

    @tailrec
    def loop(totalCost: Double): Double = {
      if (nodesWithCost.isEmpty) totalCost
      else {
        val (v, costOfAddingV): (Point, Double) = nodesWithCost.minBy { case (_, cost) => cost }
        nodesWithCost -= v
        nodesWithCost.foreach{
          case (z, costOfAddingZ) =>
            val dist: Double = calcDistance(v, z)
            if (costOfAddingZ > dist) nodesWithCost(z) = dist
        }
        loop(totalCost + costOfAddingV)
      }
    }

    loop(totalCost = 0)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrPoints: Int = reader.next().toInt
    val points: List[Point] = (for { _ <- 0 until nrPoints } yield {
      val List(x, y): List[Double] = convertToDoubleList(reader.next())
      Point(x, y)
    }).toList
    val result: Double = calcMinimalSpanningTree(points)
    println(result)
  }
}

File: OCaml/graph_algorithms/minimum_spanning_trees/cheapest_road_network.ml
open Batteries

type point = { x : int; y: int }

module PointHash = Hashtbl.Make(
    struct
        type t = point
        let equal { x = x1; y = y1 } { x = x2; y = y2 } = x1 = x2 && y1 = y2
        let hash = Hashtbl.hash
    end)


let calc_distance { x = x1; y = y1 } { x = x2; y = y2 } = hypot (float (x1 - x2)) (float (y1 - y2))


let calc_minimum_spanning_tree points =
    let nodes_with_cost = PointHash.create 1000 in
    List.iter (function node -> PointHash.add nodes_with_cost node infinity) points;
    let starting_point = List.hd points in
    PointHash.replace nodes_with_cost starting_point 0.0;
    let rec loop total_cost =
        if PointHash.is_empty nodes_with_cost then total_cost
        else
            let v, cost_of_adding_v =
                PointHash.fold (fun node cost (min_node, min_cost) ->
                    if cost < min_cost then node, cost
                    else min_node, min_cost) nodes_with_cost ({ x = 0; y = 0 }, max_float) in
            PointHash.remove nodes_with_cost v;
            let update z cost_of_adding_z =
                let dist = calc_distance v z in
                if cost_of_adding_z > dist then PointHash.replace nodes_with_cost z dist in
            PointHash.iter update nodes_with_cost;
            loop (total_cost +. cost_of_adding_v)
    in loop 0.0


let read_two_ints() = Scanf.sscanf (read_line ()) "%d %d" (fun x y-> x, y)


let read_points nr_points =
    let rec loop acc k =
        if k = nr_points then acc
        else
            let x, y = read_two_ints() in
            loop ({ x; y } :: acc) (k + 1)
    in loop [] 0


let () =
    let nr_points = read_int() in
    let points = read_points nr_points in
    let result = calc_minimum_spanning_tree points in
    print_float result;
    print_newline()
